### DISCUSSION TITLE
External point charges

### DISCUSSION THREAD

**Discussion by aizvorski** [openmopac/mopac#177] [Q&A]
**Is your feature request related to a problem? Please describe.**
I'm looking for a way to add point charges to a system.  

**Describe the solution you'd like**
I would like an option to specify the charge and coordinates of external point charges.  Something similar to Orca's `%pointcharges`,  Psi4's [`QMMM.addChargeAngstrom()`](https://psicode.org/psi4manual/1.4.0/api/psi4.driver.qmmm) or xtb's [`$embedding`](https://xtb-docs.readthedocs.io/en/latest/pcem.html).

**Describe alternatives you've considered**
I looked at the FIELD keyword but that adds a uniform electric field.  I also looked at the QMMM keyword, and it seems this needs a quantity φ(i) of the effective potential at each QM atom, rather than the charge and coordinates of the external point charges creating the potential.

I'm not sure if it is possible to produce the same effect using QMMM - would calculating sum of q_i / r_ij for each atom i and all point charges  j do what the equivalent Orca / Psi4 / xtb options do?

**Comment by godotalgorithm** [comment]
MOPAC's QM/MM interface using the [`QMMM` keyword](http://openmopac.net/manual/QMMM.html) is not completely equivalent to a QM calculation with a set of external point charges. MOPAC's interface only allows you to adjust the value of the external potential seen by each atom, but it does not allow you to specify derivative information of the external potential at each atom. In effect, MOPAC is modeling a strictly monopole/monopole interaction between MM charges and the partial charges on the QM atoms, thus the present interface is only compatible with QM/MM methods that restrict the electrostatic interactions between QM and MM regions to a monopole/monopole form. Also, if you want consistent forces, then you have to calculate the QM/MM contribution yourself because this also requires derivative information. I don't know the full history of this feature, but the present QM/MM interface in MOPAC was either requested or contributed by people actively performing QM/MM simulations, so they decided for some reason that it was best to do things this way.

If you really want point charges, then MOPAC allows you to add ["sparkles"](http://openmopac.net/manual/sparkle.html) to the list of atoms, which correspond to certain elements and ligands that are approximated as classical point charges. Unfortunately, these are only allowed to take on specific integer and half-integer charge values at the moment. It would be useful to specify sparkles with an adjustable partial charge, which should be straightforward to implement with MOPAC's existing system of atom labeling. I will consider this to be a feature request and look into implementing it at some point early next year.

**Comment by aizvorski** [comment]
@godotalgorithm Thanks!  I thought the existing QMMM keyword might work like that but I wasn't sure.  I can see how for distant charges ignoring the derivative would be a good approximation, but for very close charges it would be less good (ie the atoms immediately outside the QM region).   

Adjustable charge sparkles sounds good.  Let me know when there is stuff to test, I can compare Mopac with Orca with both running PM3.

**Comment by aizvorski** [comment]
P.S. Would there need to be some damping parameter like what xtb calls "chemical hardness" for when the orbitals overlap with the point charge?

**Comment by godotalgorithm** [comment]
Yes, MOPAC's model of sparkles is a 2-parameter model - the charge of the sparkle and an effective ionic radius (i.e. chemical hardness). Both MOPAC and xTB appear to set this radius using a Klopman-Ohno model form for the Coulomb interaction, but I'm not sure if all the details are exactly the same.

**Comment by godotalgorithm** [comment]
I've moved this from an Issue to a Discussion because it became more about the clarification of an existing feature, rather than a request for a new feature.

I also want to clarify something about the previous discussions after having read more about QM/MM methodology since then. The [standard QM/MM methodology based on semiempirical models as the QM part](https://doi.org/10.1002/jcc.540110605) actually does only prescribe monopole-monopole coupling between classical MM charges and partial atomic charges in the QM calculation. MOPAC's QMMM feature is exactly enabling the QM side of that methodology. From an ab-initio perspective, it makes more sense to couple the classical MM charges to the full electron plus nuclear charge density on the QM side of the calculation, but a lot of ab initio QM/MM calculations seem to follow an atomic partial charge prescription. Be mindful that just dropping classical point charges into a QM calculation is consistent with QM-MM coupling to the full QM charge distribution, not a set of atomic partial charges for the QM subsystem.

**Comment by garantes** [comment]
Hi, any updates here? This, as requested in the first message, would be a useful feature. The QMMM implementation is awkward and not general...

**Reply by godotalgorithm** [reply]
Not much. I made some minor adjustments to [the manual entry for the `QMMM` keyword](http://openmopac.net/manual/QMMM.html) in response to a recent request for clarifications, but no other changes have yet been made.

The existing QM/MM features were contributed to MOPAC by the developers of the [MOLARIS-XG software](https://laetro.usc.edu/molaris.html), and thus should be sufficient for its QM/MM features. Presumably, any changes to MOPAC's implementation would enable use with some other QM/MM formalism (I'm not broadly familiar with the QM/MM literature, so I don't know what variety is out there) or enable more convenient use of this established formalism with some other software.

What changes would make this feature less awkward or more general in a way that is useful? The existing implementation is based on setting the electrostatic potential at each QM atom that is generated by all the MM atoms/charges. This input is used instead of a set of MM charges because MOPAC does not have a fast Poisson solver or Coulomb summation algorithm while many MM codes do - such MM codes can generate these electrostatic potentials much faster than MOPAC can (especially if there are a lot more MM atoms than QM atoms), unless MOPAC had its own fast electrostatics. The format of the electrostatic potentials on QM atoms could change (e.g. units, file format, optional inclusion into the main input file), but it isn't clear to me what specifically is more convenient for users without further user feedback. Is there a QM/MM formalism that fully accounts for the QM-MM electrostatic coupling beyond the monopole-monopole term (and can you provide representative references)? Is there specific MM software with a QM/MM feature that you want a more convenient interface with?

Ideally, the developers of some other QM/MM-enabled MM software should provide detailed feedback for what a convenient QM/MM interface with MOPAC would be, or contribute an interface implementation themselves, just as MOLARIS-XG did.

**Reply by garantes** [reply]
Hi and thanks for the quick reply!

The QM/MM formalism is indeed a critical issue. The MOLARIS-XG approach by Luzhkov and Warshel employs a CNDO-type QM/MM electrostatic interaction, calculating only the one-electron diagonal elements (H_vv). In contrast, the standard treatment by Field and Karplus (JCC 1990, as cited above) is an NDDO-type that also includes off-diagonal interactions (H_nv).

For semiempirical QM methods, the interaction between QM cores and MM point charges is "dressed" as (ss|ss) integrals, as outlined in the same JCC 1990 paper.

Thus, the QMMM keyword in MOPAC differs from standard implementations in other codes (e.g., see initial post here).

Regarding the scaling of Coulomb summation, MM programs typically handle it using neighbor lists and cut-offs. ORCA provides an efficient implementation of this. For my purposes, I want to test Hamiltonians in MOPAC with the standard QM/MM treatment on small model systems, so a full summation of QM/MM electrostatics will do.                                    

A standard input format for partial charges (e.g., an external file containing XYZ positions and charge values) consistent with other codes would also be useful.       

As for the "monopole-monopole term", I find it unnecessary to go beyond this since most QM/MM implementations do not. That said, the terminology can be confusing due to the similarity to semiempirical ERI naming...                   

Thank you again for considering the addition of the standard QM/MM point-charge implementation into OpenMopac. I appreciate your efforts and look forward to it!

Guilherme

**Reply by godotalgorithm** [reply]
I see. When I was previously referring to the Field and Karplus paper, I didn't realize that the paper was describing 2 distinct forms of QM-MM coupling, Eq. (11) versus Eqs. (12) through (14) in the reference. Eq. (11) is what I would refer to as a "monopole-monopole" term and what the MOLARIS-XG devs contributed to MOPAC, whereas Eq. (12) is a more complete "monopole-multipole" term that more correctly describes how the QM atoms would respond to external charges.

I don't see any problems with implementing the long-range portions of Eqs. (12) and (13), but the short-ranged contact terms in Eqs. (13) and (14) are more problematic because they contain atom-specific parameters. These core-core terms only contribute a classical correction to the total energy and do not effect the electronic/QM portion of the calculation. Would it make sense to expect the external calling code to account for these classical corrections, or is that something you would want in MOPAC itself? I'd have to do some digging, but the 5 atom-specific parameters listed there are probably a subset of the 8 parameters that MOPAC currently uses in its "sparkle" model. Unless MM and QM atoms get very close together, these non-electrostatic terms will be essentially zero. The Field and Karplus paper suggests a specific choice of parameters for use with all MM atoms, but it is unclear if this prescription still makes sense and is compatible with other, more modern semiempirical models.

Also, I am not clear about how link atoms should be treated. MOPAC's current QM/MM feature does not make any distinction between link and non-link atoms. The Field and Karplus paper seems to indicate that link atoms should not see the charge of MM atoms, which would require MOPAC to account for link atoms as distinct objects with different behaviors. Are link atoms always hydrogen-like, or do there need to be different species of link atoms?

Is there any open-source software implementation of the QM/MM coupling prescribed by Field and Karplus that I could use as reference?

The minimal implementation that I will consider is to allow for input atoms labelled with the element symbol "MM", followed by a partial charge specified in the element label. There is already a [request to specify formal charges in the element labels](#40), and this would be straightforward to implement at the same time. Such a minimal implementation would ignore the classical, non-electrostatic terms. I'm still not clear about link atoms and whether or not it is sufficient to implement only a hydrogen-like link atom.

I can't make any promises about this. Unless something changes, this is the last full year that I have financial support for MOPAC development, and my top priority is creating an open-source website for MOPAC so that I can finally decommission [the old, closed-source website](http://openmopac.net). I am personally committed to the long-term support and maintenance of MOPAC, but that will likely be charity work with a very small time commitment and no new development beyond what is contributed by others.

**Reply by garantes** [reply]
Bear in mind there are two main differences between current QMMM in MOPAC (Luzhkov, JCC 1992) and the standard QM/MM in other programs (Field, JCC 1990):
1) CNDO Vs. NDDO one-electron matrix interaction with MM point-charges (eq. 4 in Luzhkov Vs. eq. 10 in Field)
2) Naked Vs. dressed QM/MM electrostatics between QC cores and MM point-charges (eq.12-14 in Field)

The exponential and Gaussian terms in eq. 13-14 in Field are the MNDO/AM1 ones, so should be handled by the QC code. Check page 709 in Field for the (5) "optimum" parameters used in MM centers.

The best open-source software implementation for QM/MM is [pDynamo3](https://github.com/pdynamo/pDynamo3/) by Field himself! :-)

No worries if it can't be implemented in MOPAC. Just a suggestion that would help using it for QM/MM comparisons.

Cheers!
